test/Quake/cse.qke

```diff
@@ -49,14 +58,84 @@ module {
+  call @device_kernel(%q11) : (!quake.veq<?>) -> ()
+  return
 }
 
+// CHECK-LABEL:   func.func @canonicalize_concat() {
+// CHECK:           %[[VAL_0:.*]] = quake.alloca !quake.ref
+// CHECK:           %[[VAL_1:.*]] = quake.concat %[[VAL_0]] : (!quake.ref) -> !quake.veq<1>
+// CHECK:           %[[VAL_2:.*]] = quake.mz %[[VAL_1]] : (!quake.veq<1>) -> !cc.stdvec<!quake.measure>
+// CHECK:           %[[VAL_3:.*]] = quake.alloca !quake.veq<1>
+// CHECK:           %[[VAL_4:.*]] = quake.mz %[[VAL_3]] : (!quake.veq<1>) -> !cc.stdvec<!quake.measure>
+// CHECK:           %[[VAL_5:.*]] = quake.alloca !quake.veq<1>
+// CHECK:           %[[VAL_6:.*]] = quake.mz %[[VAL_5]] : (!quake.veq<1>) -> !cc.stdvec<!quake.measure>
+// CHECK:           %[[VAL_7:.*]] = quake.alloca !quake.veq<2>
+// CHECK:           %[[VAL_8:.*]] = quake.relax_size %[[VAL_7]] : (!quake.veq<2>) -> !quake.veq<?>
+// CHECK:           call @device_kernel(%[[VAL_8]]) : (!quake.veq<?>) -> ()
+// CHECK:           %[[VAL_9:.*]] = quake.alloca !quake.veq<2>
+// CHECK:           %[[VAL_10:.*]] = quake.relax_size %[[VAL_9]] : (!quake.veq<2>) -> !quake.veq<?>
+// CHECK:           call @device_kernel(%[[VAL_10]]) : (!quake.veq<?>) -> ()
+// CHECK:           return
+// CHECK:         }
+
+func.func @canonicalize_multiple_concat() {
+  %q1 = quake.alloca !quake.ref
+  %p1 = quake.alloca !quake.ref
+  %q2 = quake.concat %q1, %p1 : (!quake.ref, !quake.ref) -> !quake.veq<2>
+  %b1 = quake.mz %q2 : (!quake.veq<2>) -> !cc.stdvec<!quake.measure>
+  %q3 = quake.alloca !quake.veq<1>
+  %q4 = quake.concat %q1, %p1, %q3 : (!quake.ref, !quake.ref, !quake.veq<1>) -> !quake.veq<3>
+  %b2 = quake.mz %q4 : (!quake.veq<3>) -> !cc.stdvec<!quake.measure>
+  %q5 = quake.alloca !quake.veq<1>
+  %q6 = quake.concat %q3, %q5 : (!quake.veq<1>, !quake.veq<1>) -> !quake.veq<?>
+  %b3 = quake.mz %q6 : (!quake.veq<?>) -> !cc.stdvec<!quake.measure>
+  %q7 = quake.alloca !quake.veq<2>
+  %q8 = quake.concat %q3, %q7 : (!quake.veq<1>, !quake.veq<2>) -> !quake.veq<?>
+  call @device_kernel(%q8) : (!quake.veq<?>) -> ()
+  %q9 = quake.alloca !quake.veq<2>
+  %q10 = quake.relax_size %q9 : (!quake.veq<2>) -> !quake.veq<?>
+  %q11 = quake.concat %q7, %p1, %q10 : (!quake.veq<2>, !quake.ref, !quake.veq<?>) -> !quake.veq<?>
+  call @device_kernel(%q11) : (!quake.veq<?>) -> ()
+  return
+}
+
+// CHECK-LABEL:   func.func @canonicalize_multiple_concat() {
+// CHECK:           %[[VAL_0:.*]] = quake.alloca !quake.ref
+// CHECK:           %[[VAL_1:.*]] = quake.alloca !quake.ref
+// CHECK:           %[[VAL_2:.*]] = quake.concat %[[VAL_0]], %[[VAL_1]] : (!quake.ref, !quake.ref) -> !quake.veq<2>
+// CHECK:           %[[VAL_3:.*]] = quake.mz %[[VAL_2]] : (!quake.veq<2>) -> !cc.stdvec<!quake.measure>
+// CHECK:           %[[VAL_4:.*]] = quake.alloca !quake.veq<1>
+// CHECK:           %[[VAL_5:.*]] = quake.concat %[[VAL_0]], %[[VAL_1]], %[[VAL_4]] : (!quake.ref, !quake.ref, !quake.veq<1>) -> !quake.veq<3>
```

Comment:
However, in this test, the `concat` calls are intact and order of `relax_size` is different than input. Can you add a clarifying comment as to what this test is checking?

Reply:
This test is a "negative test" to make sure the peephole isn't removing `concat` operations that cannot be removed.

It is also checking that `concat` is reified to a constant length whenever the sizes of its arguments are themselves constant. However, that reification may also imply that a `relax_size` (virtualizing the length) has to be reintroduced for type correctness.